File: src/main/java/com/empiricus/statusviajante/controller/ViagemController.java
package com.empiricus.statusviajante.controller;

import com.empiricus.statusviajante.model.ViagemModel;
import com.empiricus.statusviajante.service.ViagemService;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/viagens")
@CrossOrigin("*")
public class ViagemController {

    @Autowired
    private ViagemService viagemService;

    //Lista todas as Viagens
    @GetMapping
    public ResponseEntity<List<ViagemModel>> GetAll() {
        return ResponseEntity.ok(viagemService.buscarTodasViagens());
    }

    //Lista Viagem por ID
    @GetMapping("/{id}")
    public ResponseEntity<ViagemModel> GetById(@PathVariable Long id) {
        return viagemService.buscarViagemId(id).map(resp -> ResponseEntity.ok(resp)).orElse(ResponseEntity.notFound().build());
    }

    //Cria Viagem
    @PostMapping
    public ResponseEntity Post(@RequestBody ViagemModel viagem) throws Exception {
        try {
            return ResponseEntity.status(HttpStatus.CREATED).body(viagemService.salvarViagem(viagem));
        } catch(Exception exception){
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(exception.getMessage());
        }
    }

    @PutMapping
    public ResponseEntity put(@RequestBody ViagemModel viagem) {
        try {
            return ResponseEntity.status(HttpStatus.CREATED).body(viagemService.salvarViagem(viagem));
        } catch(Exception exception){
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(exception.getMessage());
        }
    }

    //Deleta Viagem por ID
    @DeleteMapping("/{id}")
    public void Delete(@PathVariable Long id) {
        viagemService.deletarViagemId(id);
    }

    //NOVOS ENDPOINTS
    //Lista Viagens por Usuario ID
    @GetMapping("usuario/{idUsuario}")
    public ResponseEntity<List<ViagemModel>> GetAllViagenByIdUsuario(@PathVariable Long idUsuario) {
        return ResponseEntity.ok(viagemService.buscarViagemPorUsuario(idUsuario));
    }
}

File: src/main/java/com/empiricus/statusviajante/controller/ViagemController.kt
package com.empiricus.statusviajante.controller

import com.empiricus.statusviajante.model.CadastroUsuarioModel
import com.empiricus.statusviajante.model.ViagemModel
import com.empiricus.statusviajante.service.CadastroUsuarioService
import com.empiricus.statusviajante.service.ViagemService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.util.function.Function

@RestController
@RequestMapping("/viagens")
@CrossOrigin("*")
class ViagemController {
    @Autowired
    private val viagemService: ViagemService? = null

    @Autowired
    private val cadastroUsuarioService: CadastroUsuarioService? = null

    //Lista todas as Viagens
    @GetMapping
    fun GetAll(): ResponseEntity<List<ViagemModel?>> {
        var currentUserId = cadastroUsuarioService?.getCurrentUserId()

        return ResponseEntity.ok(viagemService?.buscarViagemPorUsuario(currentUserId))
    }

    @GetMapping("/{idViagem}")
    fun GetById(@PathVariable idViagem: Long): ResponseEntity<ViagemModel> {
        var currentUserId = cadastroUsuarioService?.getCurrentUserId()

        return currentUserId?.let {
            viagemService!!.buscarViagemId(idViagem, it)
                ?.map<ResponseEntity<ViagemModel>>(Function { resp: ViagemModel? -> ResponseEntity.ok(resp) })
                ?.orElse(ResponseEntity.notFound().build())
        } ?: ResponseEntity.notFound().build()
    }

    @PostMapping
    @Throws(Exception::class)
    fun Post(@RequestBody viagem: ViagemModel?): ResponseEntity<*> {
        return try {
            var currentIdUser = cadastroUsuarioService?.getCurrentUserId()
            viagem?.usuario = CadastroUsuarioModel(currentIdUser)

            ResponseEntity.status(HttpStatus.CREATED).body(viagemService!!.salvarViagem(viagem!!))
        } catch (exception: Exception) {
            ResponseEntity.status(HttpStatus.BAD_REQUEST).body(exception.message)
        }
    }

    @PutMapping
    fun put(@RequestBody viagem: ViagemModel?): ResponseEntity<*> {

        var currentIdUser = cadastroUsuarioService?.getCurrentUserId()
        var currentViagem = viagemService?.buscarViagemId(viagem?.idViagem!!, currentIdUser!!)

        if(currentViagem?.get()?.usuario?.idUsuario?.equals(currentIdUser)!!) {
            try {
               return ResponseEntity.status(HttpStatus.CREATED).body(viagemService!!.salvarViagem(viagem!!))
            } catch (exception: Exception) {
                return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(exception.message)
            }
        } else {
            return ResponseEntity.badRequest().body("Modificação não permitida.")
        }
    }

    @DeleteMapping("/{idViagem}")
    fun delete(@PathVariable idViagem: Long?): ResponseEntity<String> {

        var currentIdUser = cadastroUsuarioService?.getCurrentUserId()
        var currentViagem = viagemService?.buscarViagemId(idViagem!!, currentIdUser!!)

        if (currentViagem != null ) {
            if (currentViagem.get().usuario?.idUsuario?.equals(currentIdUser)!!) {
                viagemService?.deletarViagemId(idViagem)
            } else {
                return ResponseEntity.badRequest().body("Viagem não pode ser deletada")
            }
        }
       return ResponseEntity.ok().body("Viagem deletada")
    }

    //NOVOS ENDPOINTS
    //Lista Viagens por Usuario ID
    @GetMapping("usuario/{idUsuario}")
    fun GetAllViagenByIdUsuario(@PathVariable idUsuario: Long?): ResponseEntity<List<ViagemModel?>?> {
        return ResponseEntity.ok(viagemService!!.buscarViagemPorUsuario(idUsuario))
    }
}